### PR TITLE
ArmVirtPkg: Reorganize crypto dependencies

### PR DESCRIPTION
**Pull request by ardbiesheuvel**
Alternative approach to enabling h/w accelerated OpenSSL on ArmVirtPkg, switching to MbedTls in cases where OpenSSL is either overkill, or should be avoided due to its dependency on floating point (32-bit ARM)

### CODE REVIEW

**Comment by ardbiesheuvel** [comment]
@os-d

File: ArmVirtPkg/ArmVirt.dsc.inc [hunk 1]
```diff
@@ -87,8 +87,8 @@
 
   # Networking Requirements
 !include NetworkPkg/NetworkLibs.dsc.inc
-!if $(NETWORK_TLS_ENABLE) == TRUE
-  TlsLib|CryptoPkg/Library/TlsLib/TlsLib.inf
+!if $(ARCH) == AARCH64
```
Comment (os-d) [inline ArmVirtPkg/ArmVirt.dsc.inc:90]: nit: Put this under LibraryClasses.AARCH64 instead of the !if?

File: ArmVirtPkg/ArmVirt.dsc.inc [hunk 2]
```diff
@@ -443,3 +439,56 @@
     <LibraryClasses>
       NULL|EmbeddedPkg/Library/PlatformHasAcpiLib/PlatformHasAcpiLib.inf
   }
+
+!if ($(NETWORK_TLS_ENABLE) == TRUE || $(SECURE_BOOT_ENABLE) == TRUE || $(TPM2_ENABLE) == TRUE) && $(ARCH) == AARCH64
```
Comment (os-d) [inline ArmVirtPkg/ArmVirt.dsc.inc:443]: nit: I can’t remember if the arch syntax works here, too, Components.AARCH64 I think does work, LibraryClasses.AARCH64.* I think is supposed to work, I’m not sure if it does.

**Review by os-d — APPROVED** [review]

**Comment by kraxel** [comment]
Hmm, this patch does a bunch of defferent changes.  Can we untangle this please?

Change 1 is switching ARM (32-bit) to mbedtls, to allow dropping the softfloat library.  No objections here.

Change 2 is switching AARCH64 to mbed tls too in some cases.  I'd very much prefer to stick to openssl here, or at least have the option to build all components against openssl.  First because mbedtls is problematic for linux distros downstream.  Second because using openssl in PEI should not be too much of a problem.  On x86 I've checked the size impact openssl has, and with PEI needing only some hash functions for TPM measurements the size impact is quite small due to LTO removing >90% of openssl.  Third because having the asm-accelerated openssl hash function for measurements should improve boot performance.

Change 3 is switching to the crypto driver.  That makes sense, but should IMO be a separate PR request.  I have an outdated patch series to switch ovmf to the crypto driver, including some changes for cryptopkg such as creating include files for the PCD-based crypto driver feature configuration.  See https://github.com/kraxel/edk2/commits/ovmf-crypto/, feel free to pick changes from there.

**Comment by kraxel** [comment]
While being at it a slightly related issue I've trapped into:  We have an RNG problem with openssl.

RngDxe supports both trng and rndr, by switching between RngLib (handling rndr) and ArmTrngLib (handling trng) at runtime.

RngLib supports rngr only, which is not supported by kvm.   OpensslLib uses RngLib -> rng not working for openssl on kvm -> breaks TlsDxe -> breaks https boot.

**Comment by ardbiesheuvel** [comment]
> Hmm, this patch does a bunch of different changes. Can we untangle this please?
> 

Fair point.

> Change 1 is switching ARM (32-bit) to mbedtls, to allow dropping the softfloat library. No objections here.
> 

Ack

> Change 2 is switching AARCH64 to mbed tls too in some cases. I'd very much prefer to stick to openssl here, or at least have the option to build all components against openssl. First because mbedtls is problematic for linux distros downstream.

Is this because of of silly FIPS/certification concerns?

> Second because using openssl in PEI should not be too much of a problem. On x86 I've checked the size impact openssl has, and with PEI needing only some hash functions for TPM measurements the size impact is quite small due to LTO removing >90% of openssl. Third because having the asm-accelerated openssl hash function for measurements should improve boot performance.
> 

Good point.

> Change 3 is switching to the crypto driver. That makes sense, but should IMO be a separate PR request. I have an outdated patch series to switch ovmf to the crypto driver, including some changes for cryptopkg such as creating include files for the PCD-based crypto driver feature configuration. See https://github.com/kraxel/edk2/commits/ovmf-crypto/, feel free to pick changes from there.

I'll take a look.

> While being at it a slightly related issue I've trapped into: We have an RNG problem with openssl.
> 
> RngDxe supports both trng and rndr, by switching between RngLib (handling rndr) and ArmTrngLib (handling trng) at runtime.
> 
> RngLib supports rngr only, which is not supported by kvm. OpensslLib uses RngLib -> rng not working for openssl on kvm -> breaks TlsDxe -> breaks https boot.

cc @pierregondois

Could we use MdePkg/Library/DxeRngLib/DxeRngLib.inf instead?

**Comment by pierregondois** [comment]
> > RngDxe supports both trng and rndr, by switching between RngLib (handling rndr) and ArmTrngLib (handling trng) at runtime.
> > RngLib supports rngr only, which is not supported by kvm. OpensslLib uses RngLib -> rng not working for openssl on kvm -> breaks TlsDxe -> breaks https boot.
> 
> cc @pierregondois
> 
> Could we use MdePkg/Library/DxeRngLib/DxeRngLib.inf instead?

IIRC RNDR is supported by kvm. I think it worked while testing it with kvmtool. Did you face some issue on that front, or am I  not remembering well ?

I agree that using DxeRngLib would be better and give much more flexibility.
Right now the RngDxe can provide:
- Depending on the RngLib:
  - CTR-DRBG/ Arm custom DRBG through RNDR instruction
  - Unsafe BaseRngTimerLib
- TRNG through firmware

The RngLib is also used at many places without safety check (i.e. checking that the implementation is safe, meaning not the Timerlib). I was briefly working on something around that.

**Comment by kraxel** [comment]
> > Change 2 is switching AARCH64 to mbed tls too in some cases. I'd very much prefer to stick to openssl here, or at least have the option to build all components against openssl. First because mbedtls is problematic for linux distros downstream.
> 
> Is this because of of silly FIPS/certification concerns?

Yes.  Distros, especially the enterprise ones, try to minimize the number of crypto libs they are using.  The certification process needed for crypto stuff is one of the reasons.

**Comment by kraxel** [comment]
> IIRC RNDR is supported by kvm. I think it worked while testing it with kvmtool. Did you face some issue on that front, or am I not remembering well ?

The behavior I see is the rndr bit in isar0 not being set, which leads to `BaseRngLibConstructor` setting `mRndrSupported` to `FALSE`.  Which in turn makes all function fail without even trying to execute the rngr instruction.  This is with qemu.

**Comment by pierregondois** [comment]
> > IIRC RNDR is supported by kvm. I think it worked while testing it with kvmtool. Did you face some issue on that front, or am I not remembering well ?
> 
> The behavior I see is the rndr bit in isar0 not being set, which leads to `BaseRngLibConstructor` setting `mRndrSupported` to `FALSE`. Which in turn makes all function fail without even trying to execute the rngr instruction. This is with qemu.

If the RNDR bit is not set, would it be possible that the emulated CPU doesn't support RNDR instruction ? Meaning that a more recent CPU being emulated would support it. `FEAT_RNG` was enabled in `Armv8.5`.

Is is possible to get the command line you are using ?

**Comment by samimujawar** [comment]
Is it possible to try the -cpu max option, please?-M virt

**Comment by ardbiesheuvel** [comment]
> Is it possible to try the qemu '-cpu max' option, please? I think you may also require '-M virt'

`-cpu max` does not work with KVM. Under KVM, there is no CPU emulation, but only virtualization of the host CPU. This means that RNDR cannot be supported in the guest if it is not implemented on the host.

Whether or not we might be able to emulate a CPU that does support RNDR is rather immaterial: emulation is a development tool with 1-2 orders of magnitude worse performance whereas virtualization is the production use case for this firmware. Note that the same applies to kvmtool, which does not implement emulation at all. In those VMs, RNDR is also entirely unsupported unless the host CPU has support for it.

**Comment by pierregondois** [comment]
> > Is it possible to try the qemu '-cpu max' option, please? I think you may also require '-M virt'
> 
> `-cpu max` does not work with KVM. Under KVM, there is no CPU emulation, but only virtualization of the host CPU. This means that RNDR cannot be supported in the guest if it is not implemented on the host.
> 
> Whether or not we might be able to emulate a CPU that does support RNDR is rather immaterial: emulation is a development tool with 1-2 orders of magnitude worse performance whereas virtualization is the production use case for this firmware. Note that the same applies to kvmtool, which does not implement emulation at all. In those VMs, RNDR is also entirely unsupported unless the host CPU has support for it.

Yes right. I was not getting the sentence below. It is not that RNDR is not supported by kvm, it's that the underlying hardware doesn't have FEAT_RNG then.

> RngLib supports rngr only, which is not supported by kvm. OpensslLib uses RngLib -> rng not working for openssl on kvm -> breaks TlsDxe -> breaks https boot.

Assuming we all agree on this, using the `DxeRngLib` is a good idea IMO

**Comment by kraxel** [comment]
> > Is it possible to try the qemu '-cpu max' option, please? I think you may also require '-M virt'
> 
> `-cpu max` does not work with KVM. Under KVM, there is no CPU emulation, but only virtualization of the host CPU. This means that RNDR cannot be supported in the guest if it is not implemented on the host.

Yes, it's kvm + `-cpu host`.  How can I figure whenever the host cpu supports rngr?

Host machines I've tries are an raspberry pi 4 and a macbook (with M2).

**Comment by pierregondois** [comment]
```
cat /proc/cpuinfo | grep Features | head -1
```
should include `rng` as a supported feature, but this has only been on models so far

**Comment by kraxel** [comment]
> ```
> cat /proc/cpuinfo | grep Features | head -1
> ```
> 
> should include `rng` as a supported feature, but this has only been on models so far

No `rng` on both machines.

**Comment by ardbiesheuvel** [comment]
> > ```
> > cat /proc/cpuinfo | grep Features | head -1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > should include `rng` as a supported feature, but this has only been on models so far
> 
> No `rng` on both machines.

Not surpising: check Marcin's ARM CPU feature table: https://gpages.juszkiewicz.com.pl/arm-socs-table/arm-socs.html, this has just started getting rolled out only on the most recent designs.

**Comment by ardbiesheuvel** [comment]
I've dropped the CryptoDxe changes for now, and focused on moving 32-bit ARM to MbedTls so we can get rid of ArmSoftFloatLib

**Review by kraxel — APPROVED** [review]

File: ArmVirtPkg/ArmVirtQemu.dsc [hunk 3]
```diff
@@ -102,16 +102,16 @@
 
 [LibraryClasses.common.PEIM]
   ArmVirtMemInfoLib|ArmVirtPkg/Library/QemuVirtMemInfoLib/QemuVirtMemInfoPeiLib.inf
-
-!if $(TPM2_ENABLE) == TRUE
```
Comment (pierregondois) [inline ArmVirtPkg/ArmVirtQemu.dsc:106]: NIT: Should the `!if $(TPM2_ENABLE) == TRUE` be kept ?

Comment (ardbiesheuvel) [reply ArmVirtPkg/ArmVirtQemu.dsc:106]: I'm not sure why I added those to begin with: there is nothing wrong with providing a library class resolution that remains unused if TPM2_ENABLE happens to be unset.

**Review by pierregondois — APPROVED** [review]

**Review by leiflindholm — APPROVED** [review]

**Pull request merged**